runtime/oti/VMHelpers.hpp

```diff
@@ -2070,57 +2089,87 @@ class VM_VMHelpers
  */
 	static VMINLINE bool
-	tryWinningConcurrentGCScan(J9VMContinuation *continuation)
+	tryWinningConcurrentGCScan(J9VMContinuation *continuation, bool isGlobalGC)
 	{
-		return J9_GC_CONTINUATION_STATE_INITIAL == VM_AtomicSupport::lockCompareExchange(&continuation->state, J9_GC_CONTINUATION_STATE_INITIAL, J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+		uintptr_t complementGCConcurrentState;
+		uintptr_t returnedState;
+		do {
+			/* preserve the concurrent GC state for the other type of GC */
+			complementGCConcurrentState = continuation->state & getGCConcurrentState(!isGlobalGC);
+			returnedState = VM_AtomicSupport::lockCompareExchange(&continuation->state, complementGCConcurrentState, complementGCConcurrentState | getGCConcurrentState(isGlobalGC));
+		/* if the other GC happened to change its concurrentGC state since us taking a snapshot of their state, we'll have to retry */
+		} while (complementGCConcurrentState != (returnedState & complementGCConcurrentState));
+
+		return (complementGCConcurrentState == returnedState);
 	}
 
+	/**
+	 * clear CONCURRENTSCANNING flag bit0:for LocalConcurrentScanning /bit1:for GlobalConcurrentScanning base on checkConcurrentState,
+	 * if all CONCURRENTSCANNING bits(bit0 and bit1) are cleared and the continuation mounting is blocked by concurrent scanning, notify it.
+	 * @param [in] checkConcurrentState can be J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN_LOCAL or J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN_GLOBAL
+	 */
 	static VMINLINE void
-	exitConcurrentGCScan(J9VMContinuation *continuation)
-	{
-		/* clear CONCURRENTSCANNING flag */
-		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
-		J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
-		if (NULL != carrierThread) {
-			omrthread_monitor_enter(carrierThread->publicFlagsMutex);
-			/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
-			omrthread_monitor_notify_all(carrierThread->publicFlagsMutex);
-			omrthread_monitor_exit(carrierThread->publicFlagsMutex);
+	exitConcurrentGCScan(J9VMContinuation *continuation, bool isGlobalGC)
+	{
+		/* clear CONCURRENTSCANNING flag bit0:LocalConcurrentScanning /bit1:GlobalConcurrentScanning */
+		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~getGCConcurrentState(isGlobalGC));
+		uintptr_t complementGCConcurrentState = oldContinuationState & getGCConcurrentState(!isGlobalGC);
+		if (!complementGCConcurrentState) {
+			J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
+			if (NULL != carrierThread) {
+				omrthread_monitor_enter(carrierThread->publicFlagsMutex);
+				/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
```

Comment:
expand this comment:

notify the waiting carrierThread that we just finished scanning and we were the only/last GC to scan it, so that it can proceed with mounting